src/uct/ib/mlx5/ib_mlx5.c

```diff
@@ -554,7 +554,9 @@ uct_ib_mlx5_devx_alloc_uar(uct_ib_mlx5_md_t *md, unsigned flags, int log_level,
     uar = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context, flags);
     if (uar == NULL) {
         sprintf(buf, "mlx5dv_devx_alloc_uar(device=%s, flags=0x%x(%s)) "
-                "failed: %m", uct_ib_device_name(&md->super.dev), flags, title);
+                "failed: %m. This may be due to insufficient UAR space. "
+                "Consider increasing PF_LOG_BAR_SIZE from the default value of 5.",
```

Comment:
1. "Consider increasing PF_LOG_BAR_SIZE using mlxconfig tool",
2. should we also check errno to have more specific error condition?
3. i think there could be other places that may fail because of the same reason, for example "mlx5dv_devx_obj_create(CQ)"

Reply:
1. added the "mlxconfig tool" to the string.
2. do you mean check "ENOMEM"?
3. added this, checking for other places

Reply:
done

Reply:
2. It looks like `mlx5dv_devx_alloc_uar` -> `_mlx5dv_devx_alloc_uar` -> `mlx5_attach_dedicated_uar` -> `mlx5_alloc_dyn_uar` -> `execute_ioctl` -> `ioctl`
it returns the `errno` but I don't know if we can know for sure that it'll set `ENOMEM` when `PF_LOG_BAR_SIZE` is insufficient. 
3. `mlx5dv_devx_obj_create(CQ)` doesn't fail because of `PF_LOG_BAR_SIZE`, in the email "
After applying the PF_LOG_BAR_SIZE parameter update".

**Regarding non-DEVX flows**
`uct_worker_tl_data_get` -> `uct_ib_mlx5_iface_get_res_domain` -> `uct_ib_mlx5_res_domain_init` -> `ibv_alloc_td` -> `mlx5_alloc_td` -> `mlx5_attach_dedicated_uar`-> `mlx5_alloc_dyn_uar` -> `execute_ioctl` -> `ioctl`
If that fails due to insufficient `PF_LOG_BAR_SIZE`, `uct_ib_mlx5_res_domain_init`  function does not fail. Instead, it returns `UCS_OK` and logs a debug message. 
The fallback mechanism uses the PD directly from the MD structure. This ensures continued operation, though with potentially reduced performance, especially for users sending a high volume of messages to different destinations. 
This is because the shared UAR may not handle the high message rate as efficiently.
I can add this if `ibv_alloc_td` fails in `uct_ib_mlx5_res_domain_init`:
`ucs_log(log_level, "Consider increasing PF_LOG_BAR_SIZE using mlxconfig tool");
`But it might not be noticeable, because entire flow doesn't fail, leading to missed logs for affected users.
Another suggestion - change the log level from `ucs_debug` to `ucs_warn`:
`ucs_warn("ibv_alloc_td() on %s failed: %m. Falling back to using the PD directly from the MD structure, which may reduce performance.", uct_ib_device_name(&md->dev))`
and then the first suggestion would be more noticeable.

Reply:
@michal-shalev  per our offline discussion,
- (2) ENOMEM - please check whether it indeed returns.
- Other flows - IMO generally we shouldn't change the behavior in this PR. Flows should not turn from failing to succeeding as a result of this PR, neither logging levels should change. Only add a suggestion for the user, in the already existing log level.

Reply:
@gleon99 I'll write a test that calls `mlx5dv_devx_alloc_uar` until it fails and checks the errno to make sure what it returns exactly.
Other flows - I'll add a suggestion.

Reply:
Ready for re-review @gleon99 @Artemy-Mellanox